baseapp/abci.go

```diff
@@ -561,6 +563,91 @@ func (app *BaseApp) createQueryContext(height int64, prove bool) (sdk.Context, e
 	return ctx, nil
 }
 
+// GetBlockRentionHeight returns the height for which all blocks below this height
+// are pruned from Tendermint. Given a commitment height and a non-zero local
+// minRetainBlocks configuration, the retentionHeight is the smallest height that
+// satisfies:
+//
+// - Unbonding (safety threshold) time: The block interval in which validators
+// can be economically punished for misbehavior. Blocks in this interval must be
+// auditable e.g. by the light client.
+//
+// - Logical store snapshot interval: The block interval at which the underlying
+// logical store database is persisted to disk, e.g. every 10000 heights. Blocks
+// since the last IAVL snapshot must be available for replay on application restart.
+//
+// - State sync snapshots: Blocks since the oldest available snapshot must be
+// available for state sync nodes to catch up (oldest because a node may be
+// restoring an old snapshot while a new snapshot was taken).
+//
+// - Local (minRetainBlocks) config: Archive nodes may want to retain more or
+// all blocks, e.g. via a local config option min-retain-blocks. There may also
+// be a need to vary retention for other nodes, e.g. sentry nodes which do not
+// need historical blocks.
+func (app *BaseApp) GetBlockRentionHeight(commitHeight int64) int64 {
+	// pruning is disabled if minRetainBlocks is zero
+	if app.minRetainBlocks == 0 {
+		return 0
+	}
+
+	min := func(x, y int64) int64 {
+		if x < y {
+			return x
+		}
+		return y
+	}
+
+	// Define retentionHeight as the minimum value that satisfies all non-zero
+	// constraints. All blocks below (commitHeight-retentionHeight) are pruned
+	// from Tendermint.
+	var retentionHeight int64
+
+	// Define the number of blocks needed to protect against misbehaving validators
+	// which allows light clients to operate safely. Note, we piggy back of the
+	// evidence parameters instead of computing an estimated nubmer of blocks based
+	// on the unbonding period and block commitment time as the two should be
+	// equivalent.
+	cp := app.GetConsensusParams(app.deliverState.ctx)
+	if cp != nil && cp.Evidence != nil && cp.Evidence.MaxAgeNumBlocks > 0 {
+		retentionHeight = commitHeight - cp.Evidence.MaxAgeNumBlocks
+	}
+
+	// Define the state pruning interval, i.e. the block interval at which the
+	// underlying logical database is persisted to disk.
+	statePruningInterval := int64(app.cms.GetPruning().Interval)
```

Comment:
I think this should use `KeepEvery`, not `Interval`. Consider the case of `KeepEvery: 1000` and `Interval: 10`. At height 10890, this will allow block pruning up to 10880, even though the last state height persisted to disk is 10000 - if the node restarts, it will try to replay from 10000, but can't since the blocks are gone.

Reply:
You're absolutely right! Can't believe I missed this. `Interval` is when the heights are actually pruned from disk -- nomenclature always confuses me...

Reply:
However, in this case `H=10890`, and `X=H-KeepEvery=9890` which isn't totally correct either. I think we need to do `X=H-(H%KeepEvery)`?

Reply:
>I think we need to do `X=H-(H%KeepEvery)`?

Yes, that's more accurate. `H-KeepEvery` would work, but has an excessive safety margin. If you go for `minNonZero` then take care when `H<KeepEvery`, since this expression results in `0` and thus would be ignored.

Reply:
> take care when `H<KeepEvery`, since this expression results in `0` and thus would be ignored.

Correct, in this case, we must be sure to return `0` as to not prune anything because that would be bad!

Reply:
Updated. @erikgrinaker could you take another look?

Reply:
Looks good!

It occurred to me that another way to handle this would be to simply return a configuration error if `minRetainBlocks` was set lower than all these other parameters - instead of magically extending it. Not really sure which UX is better.